### PR TITLE
:bug: Fix one-time-use of variables for the `ctrl` modifier

### LINKED ISSUE #1427: 🐛 MLIR - Fix one-use for modifiers in `QCO`
### Problem Statement

Currently, the modifiers (specifically `CtrlOp`) in the `QCO` dialect accept the target qubits as its arguments and the same `mlir::Value`s are then used as operands for the operation contained in the modifier. This is a theoretical violation against the one-use policy since `qubit.hasOneUse()` will return `false` and `qubit.getUses()` will also return a range with exactly two uses.

### Proposed Solution

The modifier could consume the block argument and create a new `mlir::Value` which can then be consumed by the actual operation.

Further discussion is still required on the details of this proposal and if/how this is actually possible.

### PR DESCRIPTION
**Pull request by DRovara**
## Description

With the current design of QCO's `ctrl` blocks, targets have to be specified in the `qco.ctrl` call and can then be used inside the region's body. However, this violates linear types: the corresponding variables will have two uses, one in the `qco.ctrl` signature and another one when actually using them inside the body.

To solve this issue, the `qco.ctrl` operation now defines aliases that have to be used inside the body, consistent with how, e.g., the `scf` dialect handles such situations.

For example, a `qco.ctrl` call may now look like this:

```
%controls_out, %targets_out = qco.ctrl(%0) targets (%arg0 = %1) {
  %3 = qco.h %arg0 : !qco.qubit -> !qco.qubit
  qco.yield %3
} : ({!qco.qubit}, {!qco.qubit}) -> ({!qco.qubit}, {!qco.qubit})
```

This change not only required an update to the tablegen file for QCO, but also:
- Custom parsers/printers for this assembly format.
- Updates to the conversion batterns for QC <-> QCO

Fixes #1427 

## Checklist:

- [x] The pull request only contains commits that are focused and relevant to this change.
- [x] I have added appropriate tests that cover the new/changed functionality.
- [x] I have updated the documentation to reflect these changes.
- [x] The changes follow the project's style guidelines and introduce no new warnings.
- [x] The changes are fully tested and pass the CI checks.
- [x] I have reviewed my own code changes.

### CODE REVIEW

File: mlir/lib/Dialect/QCO/IR/QCOOps.cpp [hunk 1]
```diff
@@ -21,6 +21,78 @@
...
+      // Parse "="
+      if (parser.parseEqual())
+        return failure();
+
+      // Parse "%old"
+      if (parser.parseOperand(oldOperand))
+        return failure();
+      operands.push_back(oldOperand);
+
+      Type type = qco::QubitType::get(parser.getBuilder().getContext());
```
Comment (DRovara) [inline mlir/lib/Dialect/QCO/IR/QCOOps.cpp]: I have one minor issue with the way this is implemented.

To create a new block argument, we *need* to have a type. The way this is usually done is by including the type information in the assembly format, so that i looks like this:

```
(%new = %old : !qco.qubit)
```

rather than this:

```
(%new = %old)
```

However, if we bind the type information while parsing the target aliases, we can no longer parse them  in the tablegen file with

```
`(` `{` type($controls_in) `}` ( `,` `{` type($targets_in)^ `}` )? `)`
`->`
`(` `{` type($controls_out) `}` ( `,` `{` type($targets_out)^ `}` )? `)`
```

because then, `type($targets_in)` would be bound twice.

My first thought was: no problem, then let's just remove the `type($targets_in)` binding at the end of the assembly string (so that we no longer have `: (in_types) -> (out_types)` but only `: (out_types)`. However, that does not quite work because then `type($controls_in)` is *never* bound. And in my opinion, the format `: (ctrl_in_types) -> (all_out_types)` just is not clean.

That only leaves two potential solutions:

1) Extend the custom parser so it covers the entire assembly string and not just the `(%new = %out)` part.
2) Hard-code the `!qco.qubit` type for the aliases.

Custom parsers become very big very fast, so I believe having a custom parser for the entire assembly format of the `qco.ctrl` operation would be huge.
On the other hand, we know that these aliases will always be of type `!qco.qubit` so it seems like there is no harm in this approach.

That's why I have chosen it, even though it may not be perfectly clean.

@coderabbitai what do you think about this dilemma?

File: mlir/lib/Dialect/QCO/IR/QCOOps.cpp [hunk 2]
```diff
@@ -21,6 +29,88 @@
...
+
+  // 4. Parse the Region
+  // We explicitly pass the blockArgs we just parsed so they become the entry
+  // block!
+  if (parser.parseRegion(region, blockArgs)) {
+    return failure();
+  }
+
+  return success();
+}
```
[the remark this answers is not in the document]
Comment (DRovara) [reply mlir/lib/Dialect/QCO/IR/QCOOps.cpp]: @coderabbitai I have added the verifier that checks whether numTargets <-> numBlockArgs matches.

The datatype modifier is not necessary, in my opinion.

File: mlir/lib/Dialect/QCO/IR/QCOOps.cpp [hunk 3]
```diff
@@ -21,6 +29,88 @@
...
+      printer << ", ";
+    }
+    printer.printOperand(blockArgs[i]);
+    printer << " = ";
+    printer.printOperand(targetsIn[i]);
+  }
+  printer << ") ";
+
+  printer.printRegion(region, /*printEntryBlockArgs=*/false);
+}
```
[the remark this answers is not in the document]
Comment (DRovara) [reply mlir/lib/Dialect/QCO/IR/QCOOps.cpp:114]: @coderabbitai This will now be enforced by the verifier.

File: mlir/unittests/Dialect/QCO/IR/Modifiers/test_qco_ctrl.cpp [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+  EXPECT_FALSE(region.empty());
+  auto& block = region.front();
+  EXPECT_EQ(block.getOperations().size(), 2);
+
+  EXPECT_TRUE(isa<XOp>(block.front()));
+  EXPECT_TRUE(isa<YieldOp>(block.back()));
+
+  // The template op 'xOp' still exists in the main block before ctrlOp.
+  EXPECT_TRUE(module->verify().succeeded());
+}
```
Comment (DRovara) [inline mlir/unittests/Dialect/QCO/IR/Modifiers/test_qco_ctrl.cpp:104]: @coderabbitai the way the builder is set up, it should automatically replace the usage of `q1` in the XOp inside the ctrl to the aliased variable, so I think this is okay.

File: mlir/unittests/Dialect/QCO/IR/Modifiers/test_qco_ctrl.cpp [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+  auto& block = region.front();
+
+  OpBuilder::InsertionGuard guard(builder);
+  builder.setInsertionPoint(&block.back()); // Before Yield
+  // We can insert another XOp
+  builder.create<XOp>(builder.getUnknownLoc(), block.getArgument(0));
+
+  // Should fail because body must have exactly 2 operations
+  EXPECT_TRUE(ctrlOp.verify().failed());
+}
```
[the remark this answers is not in the document]
Comment (DRovara) [reply mlir/unittests/Dialect/QCO/IR/Modifiers/test_qco_ctrl.cpp:129]: @coderabbitai this should be okay by construction

File: mlir/unittests/Dialect/QCO/IR/Modifiers/test_qco_ctrl.cpp [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+  EXPECT_EQ(block.getNumArguments(), 1); // 1 target block arg
+  EXPECT_EQ(block.getArgument(0).getType(), qType);
+
+  // Verify the XOp inside region uses block argument
+  auto innerXOp = cast<XOp>(block.front());
+  EXPECT_EQ(innerXOp.getOperand(), block.getArgument(0));
+
+  // The template op 'xOp' still exists in the main block before ctrlOp.
+  EXPECT_TRUE(mlir::verify(ctrlOp).succeeded());
+}
```
Comment (DRovara) [inline mlir/unittests/Dialect/QCO/IR/Modifiers/test_qco_ctrl.cpp:104]: @coderabbitai this is okay because the builder of the ctrl operation will fix the issue.

File: mlir/unittests/Dialect/QCO/IR/Modifiers/test_qco_ctrl.cpp [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+  bool testParse(StringRef ctrlOpAssembly) {
+    // Wrap the op in a function to provide operands
+    const std::string source =
+        (Twine("func.func @test(%q0: !qco.qubit, %q1: !qco.qubit) {\n") +
+         ctrlOpAssembly + "\n" + "  return\n" + "}")
+            .str();
+    const ScopedDiagnosticHandler diagHandler(&context);
+    // Parse should fail
+    return parseSourceString<ModuleOp>(source, &context).get() == nullptr;
+  };
```
[the remark this answers is not in the document]
Comment (burgholzer) [reply mlir/unittests/Dialect/QCO/IR/Modifiers/test_qco_ctrl.cpp]: It is actually intended to get diagnostics in the tests. so this is fine.

File: mlir/unittests/Dialect/QCO/IR/Modifiers/test_qco_ctrl.cpp [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+#include <mlir/IR/Diagnostics.h>
+#include <mlir/IR/MLIRContext.h>
+#include <mlir/IR/Operation.h>
+#include <mlir/IR/OwningOpRef.h>
+#include <mlir/IR/Value.h>
+#include <mlir/IR/ValueRange.h>
+#include <mlir/IR/Verifier.h>
+#include <mlir/Parser/Parser.h>
+#include <mlir/Support/LLVM.h>
+#include <string.h>
```
[the remark this answers is not in the document]
Comment (burgholzer) [reply mlir/unittests/Dialect/QCO/IR/Modifiers/test_qco_ctrl.cpp]: this has been resolved in the latest series of commits.

File: mlir/unittests/Dialect/QCO/IR/Modifiers/test_qco_ctrl.cpp [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+      nullptr);
+
+  // 5. Missing closing parenthesis
+  EXPECT_EQ(
+      testParse(
+          "qco.ctrl(%q0) targets (%a = %q1 { qco.yield %a } : ({!qco.qubit}, "
+          "{!qco.qubit}) -> ({!qco.qubit}, {!qco.qubit})")
+          .get(),
+      nullptr);
+}
```
[the remark this answers is not in the document]
Comment (burgholzer) [reply mlir/unittests/Dialect/QCO/IR/Modifiers/test_qco_ctrl.cpp:223]: These are just edge case tests and should be fine as they are.

File: mlir/include/mlir/Dialect/QCO/Builder/QCOProgramBuilder.h [hunk 10]
```diff
@@ -988,21 +988,21 @@ class QCOProgramBuilder final : public OpBuilder {
...
+   * %controls_out, %targets_out = qco.ctrl(%q0_in) targets(%t = %q1_in) {
+   *   %q1_res = qco.x %t : !qco.qubit -> !qco.qubit
    *   qco.yield %q1_res
    * } : ({!qco.qubit}, {!qco.qubit}) -> ({!qco.qubit}, {!qco.qubit})
    * ```
    */
   std::pair<ValueRange, ValueRange>
   ctrl(ValueRange controls, ValueRange targets,
-       const std::function<ValueRange(ValueRange)>& body);
+       const std::function<llvm::SmallVector<Value>(ValueRange)>& body);
```
[the remark this answers is not in the document]
Comment (burgholzer) [reply mlir/include/mlir/Dialect/QCO/Builder/QCOProgramBuilder.h:1007]: Switched to `llvm::function_ref<llvm::SmallVector<Value>(ValueRange)>`
Please check the current latest version whether this has been done consistently.

File: mlir/unittests/Dialect/QCO/IR/Modifiers/test_qco_ctrl.cpp [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+  xOp.erase();
+
+  // Verify structure
+  EXPECT_EQ(ctrlOp.getNumControls(), 1);
+  EXPECT_EQ(ctrlOp.getNumTargets(), 1);
+  EXPECT_EQ(ctrlOp.getResults().size(), 2); // 1 control out + 1 target out
+
+  // Verify operation
+  EXPECT_TRUE(mlir::verify(ctrlOp).succeeded());
+}
```
Comment (burgholzer) [inline mlir/unittests/Dialect/QCO/IR/Modifiers/test_qco_ctrl.cpp:104]: The test is the only one that directly uses the QCOProgramBuilder methods and, hence, does not update the tracked qubit values when applying the `ctrl` operation.
This test should be fine as-is.

File: mlir/unittests/Dialect/QCO/IR/Modifiers/test_qco_ctrl.cpp [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
```
[the remark this answers is not in the document]
Comment (burgholzer) [reply mlir/unittests/Dialect/QCO/IR/Modifiers/test_qco_ctrl.cpp:104]: See the reasoning above for why this is as-intended.

**Review by burgholzer — APPROVED** [review]
@DRovara thanks for pushing this.
I made some further changes to this that mainly do two things:
- make the `QCOProgramBuilder` work (including a signature change that was only discovered to be necessary through this PR)
- optimize various bits and pieces throughout to use LLVM/MLIR concepts wherever reasonable and to avoid redundant computation.

I am happy with this PR now, and so is the Rabbit.
If you are happy as well, feel free to merge.

**Comment by DRovara** [comment]
> @DRovara thanks for pushing this. I made some further changes to this that mainly do two things:
> 
> * make the `QCOProgramBuilder` work (including a signature change that was only discovered to be necessary through this PR)
> * optimize various bits and pieces throughout to use LLVM/MLIR concepts wherever reasonable and to avoid redundant computation.
> 
> I am happy with this PR now, and so is the Rabbit. If you are happy as well, feel free to merge.

Thanks for the changes! I guess the fact that CodeCov is not fully happy can be fixed once we actually finish up the full test system. The main functionalities I introduced here should all be tested, at least.
